### PR TITLE
Ft historical device measurements

### PR DESCRIPTION
**Pull request by NoahNsimbe**
# Airqo and Kcca historical device measurements

**_WHAT DOES THIS PR DO?_**

- [x] Extracts historical device measurements from KCCA and Things-speak
- [x] Transforms the data to match the format required by Netmanager
- [x] Adds calibrated values for the Airqo device measurements
- [x] Loads the data into Netmanager

**_WHAT JIRA STORY/TASK IS RELATED TO THIS PR?_**
[PLAT-432](https://airqoteam.atlassian.net/browse/PLAT-431?atlOrigin=eyJpIjoiZDJmOTUyOTJkYTU3NGMxMDkwZWMwMTAzOTE0YWI0MGUiLCJwIjoiaiJ9)

**_WHAT IS THE LINK TO THE PR BRANCH_**
[historical device measurements](https://github.com/airqo-platform/AirQo-api/tree/ft-historical-device-measurements)

**_HOW DO I TEST OUT THIS PR?_**

First start the luigi central shdeular. Run the following commands
> `cd AirQo-api/src/data-mgt/python/batch-jobs`
> `docker build -t luigi-central-schedular .`
> `docker run -d -p 8082:8082 luigi-central-schedular`

Browse http://localhost:8082 to visualize luigi 

To test for  KCCA 
>  `cd AirQo-api/src/data-mgt/python/batch-jobs/kcca`
>
>  Follow instructions in the Readme.md

To test  for AIRQO
>  `cd AirQo-api/src/data-mgt/python/batch-jobs/airqo`
>
>   Follow instructions in the Readme.md

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**
N/A

**_ARE THERE ANY RELATED PRs?_**
N/A

### CODE REVIEW

File: src/data-mgt/python/batch-jobs/airqo/batch_fetch_airqo_data.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                    "speed": {"value": check_float(row["speed"])},
+                    "satellites": {"value": check_float(row["satellites"])},
+                    "hdop": {"value": check_float(row["hdop"])},
+                    "internalTemperature": {"value": check_float(row["internalTemperature"])},
+                    "internalHumidity": {"value": check_float(row["internalHumidity"])},
+                })
+
+                details.append(data)
+
+                thread = Thread(target=single_component_insertion, args=(data, "airqo"))
```
Comment (Baalmart) [inline src/data-mgt/python/batch-jobs/airqo/batch_fetch_airqo_data.py:133]: We might need to review the naming to reflect reality, are we still doing single component insertion?

**Review by Baalmart — COMMENTED** [review]
Just been reviewing, great stuff @NoahNsimbe , brilliant!

- Just curious, where is the feeds url(FEEDS_BASE_URL)  being used in the recent changes?
- Are we also doing streaming here? Or is this just about batch processing of historical records?
- On a side note, the FEEDS_BASE_URL does caching so its cache time (cache validity period) of one hour needs to be reviewed accordingly in case it is being utilised anywhere.
- Could we also share the rest of the required files (secret file.json) via Slack channel for devs?
- At the end of building the airqo-batch image, I obtain the following error, so this makes it impossible for me to proceed with the tests. Able to advise?
```
-ERROR: Could not find a version that satisfies the requirement cffi==1.14.5
ERROR: No matching distribution found for cffi==1.14.5
```

**Review by sserurich — COMMENTED** [review]
both the jobs  for kcca & airqo seemed to run well, however airqo job one was giving me errors msg on the console.. not sure what that means, 
image.  next step will be to check & verify that the imported data matches the  one from the source...

**Review by Baalmart — COMMENTED** [review]
Hi @NoahNsimbe 
Just sharing notes from the most recent PR review meeting:

- We reviewed the deployment aspects (k8s) of the jobs and realised the need to have it done on the server since it is a long running job for AirQo. @danielogen 
- There is some weird behaviour of duplicate days within the same collection, the compound indices are misbehaving. @Baalmart 
- Review the visualisation aspects of the tool (Luigi) so that we can get more insights into what is happening @NoahNsimbe  
- As for the errors that happen during the running of the job, what happens to that data? Details of those errors are in the screenshots @NoahNsimbe 
- There is also a need to compare values of ThingSpeak/BigQuery with those in the Events collection just to be certain of things. @ALL and @sserurich 
- Possibility of adding time/date flags for the AirQo batch job? Otherwise, not urgent, just a nice to have if possible @NoahNsimbe

**Review by ja-odur — COMMENTED** [review]
@NoahNsimbe @Baalmart I think it's better for this job to push directly to the events collection (mongodb). This might required more work to be done but It removes the in-between `device-registry` micro-service, thus reducing on possible points of failure and possible improve on the job's speed. We want our jobs to be independent too. This is just a though.

**Comment by Baalmart** [comment]
> @NoahNsimbe @Baalmart I think it's better for this job to push directly to the events collection (mongodb). This might required more work to be done but It removes the in-between `device-registry` micro-service, thus reducing on possible points of failure and possible improve on the job's speed. We want our jobs to be independent too. This is just a though.

Sure @ja-odur , this is a very good thought/suggestion. We just need to have review it further by considering our system architecture that involves **multitenancy** and **microservices.** 

On the side note, I do not see any significant difference if the logic behind the interface (device registry endpoint) is transferred to the actual job.

A few questions/thoughts around this thought:

- When we transfer the logic from the interface (endpoint) to the job, how do we plan to handle reusability of that logic?
- How do we plan to deal with multitenancy within the job.
- How will the transfer of that logic from the interface to the job improve performance? Can't that performance be improved from the interface itself?
- If the logic of the job is transferred to the job, will it consider the DB schema design we already have in place? Or it will have to use its own design?
- What will be the scope of the job? Considerations: extract, transform and load? With the load also having more logic to enforce more DB/schema design considerations?
- How will the inclusion of the schema design enforcement logic (device registry interface) in the job have on the overall performance of the job?

**Review by ja-odur — COMMENTED** [review]
The kcca batch process job worked fine for me although the dates being inserted into the events collections dont match the range specified in the `env.kcca.batch` env file.

I tried running the `airqo` batch process job and got an error that looks likes a `JsonDecode Error`
image

**Comment by NoahNsimbe** [comment]
> The kcca batch process job worked fine for me although the dates being inserted into the events collections dont match the range specified in the `env.kcca.batch` env file.
> 
> I tried running the `airqo` batch process job and got an error that looks likes a `JsonDecode Error`
> image

**Pull request merged**